packages/js/product-editor/src/blocks/shipping-fee/edit.tsx

```diff
@@ -0,0 +1,194 @@
+import {
+	EXPERIMENTAL_PRODUCT_SHIPPING_CLASSES_STORE_NAME,
+	ProductShippingClass,
+} from '@woocommerce/data';
+import { getNewPath } from '@woocommerce/navigation';
+import { recordEvent } from '@woocommerce/tracks';
+import { useBlockProps } from '@wordpress/block-editor';
+import {
+	BaseControl,
+	RadioControl,
+	SelectControl,
+} from '@wordpress/components';
+import { useInstanceId } from '@wordpress/compose';
+import { useEntityProp } from '@wordpress/core-data';
+import { useSelect } from '@wordpress/data';
+import {
+	Fragment,
+	createElement,
+	createInterpolateElement,
+	useEffect,
+	useState,
+} from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import { ShippingFeeBlockAttributes } from './types';
+import { useValidation } from '../../hooks/use-validation';
+
+const FOLLOW_CLASS_OPTION_VALUE = 'follow_class';
+const FREE_SHIPPING_OPTION_VALUE = 'free_shipping';
+
+const options = [
+	{
+		label: __( 'Follow class', 'woocommerce' ),
+		value: FOLLOW_CLASS_OPTION_VALUE,
+	},
+	{
+		label: __( 'Free shipping', 'woocommerce' ),
+		value: FREE_SHIPPING_OPTION_VALUE,
+	},
+];
+
+export function Edit( {
+	attributes,
+}: BlockEditProps< ShippingFeeBlockAttributes > ) {
+	const { title } = attributes;
+
+	const blockProps = useBlockProps( {
+		className: 'wp-block-woocommerce-product-shipping-fee-fields',
+	} );
+
+	const [ option, setOption ] = useState< string >(
+		FREE_SHIPPING_OPTION_VALUE
+	);
+
+	const [ shippingClass, setShippingClass ] = useEntityProp< string >(
+		'postType',
+		'product',
+		'shipping_class'
+	);
+
+	const { shippingClasses } = useSelect( ( select ) => {
+		const { getProductShippingClasses } = select(
+			EXPERIMENTAL_PRODUCT_SHIPPING_CLASSES_STORE_NAME
+		);
+		return {
+			shippingClasses:
+				getProductShippingClasses< ProductShippingClass[] >() ?? [],
+		};
+	}, [] );
+
+	const shippingClassControlId = useInstanceId( BaseControl ) as string;
+
+	const isShippingClassValid = useValidation(
+		'product/shipping_class',
+		function shippingClassValidator() {
+			if ( option === FOLLOW_CLASS_OPTION_VALUE && ! shippingClass ) {
+				return false;
+			}
+			return true;
+		}
+	);
+
+	function handleOptionChange( value: string ) {
+		setOption( value );
+
+		if ( value === FOLLOW_CLASS_OPTION_VALUE ) {
+			const [ firstShippingClass ] = shippingClasses;
+			setShippingClass( firstShippingClass?.slug ?? '' );
+		} else {
+			setShippingClass( '' );
+		}
+	}
+
+	useEffect( () => {
+		if ( shippingClass === '' ) {
+			setOption( FREE_SHIPPING_OPTION_VALUE );
+		} else if ( shippingClass ) {
+			setOption( FOLLOW_CLASS_OPTION_VALUE );
+		}
+	}, [ shippingClass ] );
+
+	return (
+		<div { ...blockProps }>
+			<div className="wp-block-columns">
+				<div className="wp-block-column">
+					<RadioControl
+						label={
+							<span className="wp-block-woocommerce-product-radio__title">
```

Comment:
So that we reduce the possibility of introducing breaking changes in the future, could we use the actual block radio component here?

The pattern I've seen in GB and WC Blocks is to extract [this portion](https://github.com/woocommerce/woocommerce/blob/add/37406/packages/js/product-editor/src/components/radio/edit.tsx#L25-L44) into a separate `block.tsx` file so that we can reuse in both the radio control's `Edit` component and here.

Reply:
I would like to suggest to not reuse a part of another block if that part is within the same block folder. I think blocks should be reused only via templates and inner blocks. If we want to have shared parts then we should create components outside of each block that are gonna be using the shared parts.

Reply:
done -> https://github.com/woocommerce/woocommerce/pull/37642/commits/ae58d287c1a1dba6e0a580d71d23a2dcaf189bbf